plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionManager.java

```diff
@@ -98,27 +99,39 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
         if (partitionColumns.isEmpty()) {
             return new HivePartitionResult(
                     partitionColumns,
+                    ImmutableList.of(UNPARTITIONED_ID),
                     ImmutableList.of(new HivePartition(tableName)),
                     compactEffectivePredicate,
                     effectivePredicate,
                     TupleDomain.all(),
                     hiveBucketHandle,
-                    bucketFilter);
+                    bucketFilter,
+                    true);
         }
 
         List<Type> partitionTypes = partitionColumns.stream()
                 .map(HiveColumnHandle::getType)
                 .collect(toList());
 
         Iterable<HivePartition> partitionsIterable;
+        List<String> partitionNames;
         Predicate<Map<ColumnHandle, NullableValue>> predicate = constraint.predicate().orElse(value -> true);
         if (hiveTableHandle.getPartitions().isPresent()) {
-            partitionsIterable = hiveTableHandle.getPartitions().get().stream()
+            List<HivePartition> partitions = hiveTableHandle.getPartitions().get().stream()
                     .filter(partition -> partitionMatches(partitionColumns, effectivePredicate, predicate, partition))
                     .collect(toImmutableList());
+            partitionsIterable = partitions;
+            partitionNames = partitions.stream()
+                    .map(HivePartition::getPartitionId)
+                    .collect(toImmutableList());
         }
         else {
-            List<String> partitionNames = getFilteredPartitionNames(metastore, identity, tableName, partitionColumns, compactEffectivePredicate);
+            if (hiveTableHandle.getPartitionNames().isPresent()) {
+                partitionNames = hiveTableHandle.getPartitionNames().get();
```

Comment:
It's odd that we don't need to filter partition names while we had to do `.filter(partition -> partitionMatches(partitionColumns, effectivePredicate, predicate, partition))` above.

Comment would be nice

Reply:
ping, I still don't understand why we don't have to filter here.
Why there can be partition names and not `getPartitions`?

Reply:
`partitionMatches` effectively converts the partition names from `String` to `HivePartitionInformation` and then tries to match with the filter predicate. 

>  Why there can be partition names and not getPartitions?
The contract is that a table can either load a partition details (if it is less than a threshold) or could maintain it as raw `String` if it crosses a threshold. We don't have an intermediate state here

Reply:
> partitionMatches effectively converts the partition names from String to HivePartitionInformation and then tries to match with the filter predicate.

So in this branch:
```
partitionNames = hiveTableHandle.getPartitionNames().get();
```
we don't do any filtering. Does it mean we return excessive `partitionNames`?

Reply:
> we don't do any filtering. 

We do kind of partitial filterting, like if we have a TupleDomain (with Domain on Partition columns) then the filtering would be applied at a metastore layer but we won't perform `partitionMatches` (and materializing it into `HivePartition`) 

> Does it mean we return excessive partitionNames?

Since we dont invoke `partitionMatches` there is a chance that we could return excessive partition names.

Reply:
> Since we dont invoke partitionMatches there is a chance that we could return excessive partition names.

That doesn't seem correct, does it?

Reply:
I think in this case we don't have specify the `enforcedTupleDomain` as `TupleDomain` so that the filter expression is not lost. It will be applied during the next `applyFilter` optimizer. WDYT ?

Reply:
> I think in this case we don't have specify the enforcedTupleDomain as TupleDomain so that the filter expression is not lost. It will be applied during the next applyFilter optimizer. WDYT ?

Depends on contract. If we say `getPartitions` returns partitions that match `constrant`, then it should be the case